collect_app/src/main/java/org/odk/collect/android/tasks/ChosenFileSaver.java

```diff
@@ -0,0 +1,105 @@
+package org.odk.collect.android.tasks;
+
+import android.content.Context;
+import android.net.Uri;
+import android.provider.MediaStore;
+
+import org.odk.collect.android.R;
+import org.odk.collect.android.application.Collect;
+import org.odk.collect.android.dao.helpers.ContentResolverHelper;
+import org.odk.collect.android.exception.GDriveConnectionException;
+import org.odk.collect.android.logic.FormController;
+import org.odk.collect.android.utilities.FileUtils;
+import org.odk.collect.android.utilities.MediaUtils;
+
+import java.io.File;
+import java.lang.ref.WeakReference;
+
+import io.reactivex.Observable;
+import timber.log.Timber;
+
+public class ChosenFileSaver {
+
+    private final WeakReference weakReferenceContext;
+    private final Uri selectedFile;
+    private File newFile;
+    private int errorMessageRes;
+
+    private ChosenFileSaver(Context context, Uri selectedFile) {
+        weakReferenceContext = new WeakReference<>(context);
+        this.selectedFile = selectedFile;
+    }
+
+    public static Observable<SaveResult> createObservable(Context context, Uri selectedFile) {
+        return Observable.create(emitter -> {
```

Comment:
So with creating an observable that's supposed to do asynchronous work can be a bit tricky! Using `create` doesn't accomplish that task because create is made for synchronous work and it won't handle waiting on the result well so you have to switch to something like `Observable.defer' or `Observable.callable`. Most of the `RxJava` community has moved to `callable` because it makes writing this code concise since `callable` handles `Exceptions` really smoothly. You will see it's use on my branch.

Reply:
But this only gets called when the observable is subscribed, right? That means we can still attach the `observeOn` and `subscribeOn` to change the threads before the actual method call happens. Did I get it wrong? What you said makes sense when the Observable is getting creating instantly as in case of `Observable.just()`

What do you think? I still like your approach better but thought to clear up my doubts first